### PR TITLE
Use npms.io API to fetch infos

### LINKED ISSUE #50: Proposal: Use a external API instead of fetch and parse data from multiple service
Hi,

Currently we make some HTTP requests from multiple service (npm and Github) and we encounter the problem of rate limit on Github due to existing more than 5000 yeoman-generator.

I would start a PR to use [npms.io](https://npms.io/) to replace this process and resolve issue.

What do you think about that?

---

[Example search call API](https://api.npms.io/v2/search?q=keywords%3Ayeoman-generator)

[Example package call API](https://api.npms.io/v2/package/generator-fountain-webapp)

### PR DESCRIPTION
**Pull request by zckrs**
Close #50

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -5,14 +5,14 @@ const log = require('./logger');
 
 module.exports = {
   log: log,
-  update: (keyword, limit) => {
+  update: keyword => {
     if (typeof keyword !== 'string') {
       let msg = 'Index: Keyword is required';
       log.error(msg);
-      throw Error(msg);
+      throw new Error(msg);
```
Comment (eddiemonge) [inline src/index.js]: `new` is not necessary

File: src/merge-info.js [hunk 2]
```diff
@@ -1,43 +0,0 @@
...
-    updated: npm.updated
-  };
-};
-
-function cleanupDescription(str) {
-  if (!str) {
-    return '';
-  }
-
-  str = str.trim()
```
Comment (eddiemonge) [inline src/merge-info.js]: this part should be left in unless npm.io somehow already follows these rules, which I doubt

File: src/npms-package-info.js [hunk 3]
```diff
@@ -0,0 +1,65 @@
+'use strict';
+
+const moment = require('moment');
+const got = require('got');
+const log = require('./logger');
+
+module.exports = list => {
+  log.info('npmsInfo: Fetching info for %s packages', list.length);
+  let arrayPackages = [];
```
Comment (eddiemonge) [inline src/npms-package-info.js]: this seems to do a lot more work than 

```
return Promise.all(list.map(pkg => createPromiseUpdatePage(pkg)
.then(pkgs => {		
  pkgs = pkgs.filter(pkg => Boolean(pkg));		
  log.info('npmInfo: Fetched info for %s valid packages', pkgs.length);		
  return pkgs;		
});
```

and return the `formattedPkg` in `createPromiseUpdatePage`

File: src/index.js (hunk 1 above)
Comment (zckrs) [inline src/index.js]: XO linter says

https://github.com/sindresorhus/eslint-plugin-unicorn/blob/master/rules/throw-new-error.js#L21

File: src/npms-package-info.js (hunk 3 above)
Comment (zckrs) [inline src/npms-package-info.js]: See https://github.com/npms-io/npms-api/issues/52

File: src/index.js (hunk 1 above)
Comment (SBoudrias) [inline src/index.js]: Yeah, I'd keep new for best practices. I don't think it matters on Node, but browser have very inconsistent error traces if you user new or not.

**Comment by SBoudrias** [comment]
What is the rate limiting situation of npmsio?

**Comment by zckrs** [comment]
Except Cloudfare DDOS protection I didn't find rate limit

File: src/merge-info.js (hunk 2 above)
Comment (zckrs) [inline src/merge-info.js]: Done

File: src/npms-package-info.js (hunk 3 above)
Comment (zckrs) [inline src/npms-package-info.js]: Now I use [p-map](https://github.com/sindresorhus/p-map)

File: src/npms-package-info.js [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+  return pMap(list, fetchInfos, {concurrency: 2})
+    .then(pkgs => {
+      pkgs = pkgs.filter(pkg => Boolean(pkg));
+      log.info('npmInfo: Fetched info for %s valid packages', pkgs.length);
+
+      return pkgs;
+    })
+    .catch(err => {
+      console.error(err);
```
Comment (sindresorhus) [inline src/npms-package-info.js]: This should either use the logger or preferably just not `.catch()` and let the error bubble up to the next `.catch()`.

**Comment by sindresorhus** [comment]
// @satazor

**Comment by satazor** [comment]
You may use the multi-get endpoint to fetch several packages, instead of several in parallel. See: https://api-docs.npms.io/#api-Package-GetMultiPackageInfo

File: src/npms-package-info.js [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+const moment = require('moment');
+const got = require('got');
+const pMap = require('p-map');
+const log = require('./logger');
+
+module.exports = list => {
+  log.info('npmsInfo: Fetching info for %s packages', list.length);
+
+  return pMap(list, fetchInfos, {concurrency: 2})
```
Comment (zckrs) [inline src/npms-package-info.js]: FYI currently a run take 5 min.

**Comment by zckrs** [comment]
Thank you @satazor 

I just pushed a commit:
* to use GetMultiPackageInfo from npms.io
* to remove concurrency options in p-map (POST requests seems not catch as DDOS)


Conclusion: data are fetched in ~10 seconds for +5000 packages.

**Comment by eddiemonge** [comment]
I'll do some additional testing on it this week but seems pretty good so far. Thanks!

**Comment by zckrs** [comment]
@eddiemonge do you find time to test ? 

http://yeoman.io/generators/ stop to work due a empty https://storage.googleapis.com/generators.yeoman.io/cache.json

**Comment by diegohaz** [comment]
Any updates on this?

**Comment by zckrs** [comment]
Close in favor of https://github.com/yeoman/yeoman-generator-list/pull/52

**Pull request closed without merging**